### PR TITLE
Add option to query and upgrade cluster (v19.6.1)

### PR DESCRIPTION
**Pull request by brtkwr**
Also adds hosts in the cluster to a dynamic inventory

### CODE REVIEW

File: tasks/upgrade.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+---
```
Comment (markgoddard) [inline tasks/upgrade.yml:1]: Is there actually any overlap between these two tasks? Would it make more sense to have a separate upgrade role?

**Comment by brtkwr** [comment]
The overlap is that the upgrade role can use the dynamic inventory generated from the query. I thought about a separate role but seemed like an overkill for something that is so strongly coupled with Magnum deployed Kubernetes.

**Comment by markgoddard** [comment]
> The overlap is that the upgrade role can use the dynamic inventory generated from the query. I thought about a separate role but seemed like an overkill for something that is so strongly coupled with Magnum deployed Kubernetes.

Ok, I'll leave it up to you. It would be possible to just query using this role, then use the generated inventory with a separate role.

**Comment by brtkwr** [comment]
@markgoddard, it is true that it would be possible. I feel like like it would be nice to have all the tools for the same job in the same toolbox 😛 🛠

**Comment by brtkwr** [comment]
Besides, this is a temporary workaround as upgrades will be supported natively by the magnum api in the near future.

File: README.md [hunk 2]
```diff
@@ -16,11 +16,11 @@ Requirements
 Role Variables
 --------------
 
-`os_container_infra_cloud` is the name of the cloud inside cloud.yaml.
+`os_container_infra_cloud` is the name of the cloud inside cloud.yaml (optional).
```
Comment (oneswig) [inline README.md]: isn't it `clouds.yaml`?  Consider also giving the full pathname

File: README.md [hunk 3]
```diff
@@ -47,6 +46,12 @@ which all cluster hosts and localhost will be added. This can be useful if
 there is a single group that represents an environment such as
 development/staging/production.
 
+`os_container_infra_coe` is container orchestration engine to use. Valid
+options are `kubernetes` or `swarm`.
+
+`os_container_infra_k8s_version` is the Kubernetes version to use when
+`os_container_infra_state = upgrade`.
```
Comment (oneswig) [inline README.md:53]: Do we not get to choose a version when we are installing?

File: README.md [hunk 4]
```diff
@@ -92,36 +92,25 @@ To authenticate using information passed via playbook:
           user_domain_name: Default
           project_domain_name: Default
           region_name: RegionOne
-        # ...container specific variables
-    ...
 
 To authenticate using the information stored in `.config/openstack/clouds.yaml`
```
Comment (oneswig) [inline README.md]: Or presumably `/etc/openstack/clouds.yaml`

File: README.md [hunk 5]
```diff
@@ -136,6 +125,16 @@ To filter these warnings:
...
 
+Known Issues
+------------
+
+The `template` Ansible modules writes an inventory file and it may complain
+about missing `libselinux-python` which is already installed is most Linux
+distros. If thats the case, simply create a symlink to the selinux directory in
+your virtual environment:
+
+    ln -s /usr/lib64/python2.7/site-packages/selinux/ venv/lib64/python2.7/site-packages/
```
Comment (oneswig) [inline README.md:137]: What about the other workaround, of creating a virtualenv with `--system-site-packages`

File: tasks/main.yml [hunk 6]
```diff
@@ -1,47 +1,18 @@
...
-    auth: "{{ os_container_infra_auth or omit }}"
-    cloud: "{{ os_container_infra_cloud or omit }}"
-    state: "{{ os_container_infra_state }}"
-    cluster_name: "{{ os_container_infra_cluster_name }}"
-    cluster_template_name: "{{ os_container_infra_cluster_template_name }}"
-    master_count: "{{ os_container_infra_master_count }}"
-    node_count: "{{ os_container_infra_node_count }}"
-    keypair: "{{ os_container_infra_keypair }}"
+- name: os_container_infra_state != 'upgrade'
+  include_tasks: not-upgrade.yml
```
Comment (oneswig) [inline tasks/main.yml]: Can you think of a clearer name than `not-upgrade`?

**Review by oneswig — CHANGES_REQUESTED** [review]
not-upgrade might be the logical case but does it make sense describing the tasks out of context.  Couple of other minor questions.

File: README.md (hunk 5 above)
Comment (brtkwr) [reply README.md:137]: Done

File: README.md (hunk 4 above)
Comment (brtkwr) [reply README.md]: Added

File: README.md (hunk 3 above)
Comment (brtkwr) [reply README.md:53]: Yep, you can

File: tasks/not-upgrade.yml [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    master_count: "{{ os_container_infra_master_group.count | int }}"
+    node_count: "{{ os_container_infra_worker_group.count | int }}"
+    keypair: "{{ os_container_infra_keypair }}"
+
+- block:
+
+    - name: Get stack facts and load them into stack_facts
+      os_stack_resources:
+        auth_type: "{{ os_container_infra_auth_type or omit }}"
+        auth: "{{ os_container_infra_auth or omit }}"
```
Comment (markgoddard) [inline tasks/not-upgrade.yml]: No support for clouds.yaml?

File: tasks/not-upgrade.yml [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        name: "{{ container_infra.stack_id }}"
+        max_depth: 2
+        filters:
+          resource_type: 'OS::Nova::Server'
+
+    - name: Attach interfaces to servers
+      os_server_interface:
+        auth_type: "{{ os_container_infra_auth_type or omit }}"
+        auth: "{{ os_container_infra_auth or omit }}"
```
Comment (markgoddard) [inline tasks/not-upgrade.yml]: clouds.yaml?

File: README.md (hunk 5 above)
Comment (brtkwr) [reply README.md:137]: Added

File: tasks/main.yml (hunk 6 above)
Comment (brtkwr) [reply tasks/main.yml]: present-or-query.yml

**Review by markgoddard — COMMENTED** [review]
There's too much in this PR again, it's hurting my head.

File: tasks/not-upgrade.yml (hunk 7 above)
Comment (brtkwr) [reply tasks/not-upgrade.yml]: There is

File: defaults/main.yml [hunk 9]
```diff
@@ -1,54 +1,67 @@
...
 os_container_infra_node_count: 1
+os_container_infra_external_interface:
 
-# Type of cluster
-os_container_infra_coe: "kubernetes" # can also be swarm-mode
-
-# Kubernetes related
-os_container_infra_k8s_version: "v1.13.4"
-os_container_infra_k8s_master_kubelet: false
+# DO NOT CHANGE
```
Comment (markgoddard) [inline defaults/main.yml:45]: For future reference, constants can go in a vars/main.yml file.

**Review by markgoddard — APPROVED** [review]
Great, thanks!

**Pull request merged**